### ISSUE TITLE
Qpid Proton: Investigate options to handle subscribers going away

### ISSUE THREAD

**Issue by adriennea** [OpenMAMA/OpenMAMA#65]
Replacing [BZ-41] (http://bugs.openmama.org/show_bug.cgi?id=41)

When running a mamapublisher / mamasubscriber test, after a subscriber connects and then disconnects, you will receive the following:

Publishing message 16 to MAMA_TOPIC.	msg=({{MdMsgStatus,2,0},{MdSeqNum,10,16},{PublisherTopic,10002,MAMA_TOPIC}}
Publishing message 17 to MAMA_TOPIC.ERROR amqp:connection:framing-error connection aborted
[0xb5809090:0] ERROR[-2] connection aborted
CONNECTION ERROR connection aborted
	msg=({{MdMsgStatus,2,0},{MdSeqNum,10,17},{PublisherTopic,10002,MAMA_TOPIC}}
Publishing message 18 to MAMA_TOPIC.ERROR amqp:connection:framing-error connection aborted
[0x9c8aff8:0] ERROR[-2] connection aborted
	msg=({{MdMsgStatus,2,0},{MdSeqNum,10,18},{PublisherTopic,10002,MAMA_TOPIC}}
Publishing message 19 to MAMA_TOPIC.CONNECTION ERROR connection aborted
CONNECTION ERROR connection aborted
	msg=({{MdMsgStatus,2,0},{MdSeqNum,10,19},{PublisherTopic,10002,MAMA_TOPIC}}
Publishing message 20 to MAMA_TOPIC.CONNECTION ERROR connection aborted
	msg=({{MdMsgStatus,2,0},{MdSeqNum,10,20},{PublisherTopic,10002,MAMA_TOPIC}}
Publishing message 21 to MAMA_TOPIC.CONNECTION ERROR connection aborted

This is by design because there was no known way with the qpid proton interface to tell whether a network event has occurred (in which case publish should keep trying) or whether the subscriber simply went away. This ticket is to determine if there is a better way to handle this, possibly with some sort of timeout / cleanup on connections which have gone away for whatever reason.

**Comment by adriennea** [comment]
Hi,

I am building an application that uses the mama dq/advanced publisher and the qpid bridge.

I am seeing the same behavior in my system, when the client disconnects. Not a huge issue for a few subscriptions but if I am subscribing to > 10k symbols then disconnect this causes the publisher to slow to a crawl and I loose incoming data.

Has there been any progress made or workarounds available?

Details:
C++ g++ Linux rhel6 qpid-0.7

Many thanks

James

**Comment by adriennea** [comment]
Hi James,

Honestly, this was just one of those annoying things that we found when we were testing it - it wasn't until it was raised on the mailing list a few days ago that we realized that this was actually causing real problems (http://lists.openmama.org/pipermail/openmama-dev/2015-June/001513.html). 

Have you considered a broker based approach for your solution if you have multiple clients? We have a test branch feature-qpid-broker which contains a broker based qpid proton implementation which should make these sorts of problems go away (the broker which is specifically designed for handling fan in / fan out should then take care of issues like this).

Cheers,
Frank

**Comment by adriennea** [comment]
Hi Frank,

Thanks for the swift response.

If possible I would like to avoid the broker approach for our high volume internal systems, we still have distant memories of the tibco brokers taking out boxes on heavy market days. Point to point (or point to many points) while not being as efficient removes the single point of failure for us.

That being said I will try the broker setup out at some point, things may have changed a lot.

I guess the good thing about mama is option to change to another transport in the future anyway. Maybe a zmq bridge at some point.
 
cheers

James

**Comment by adriennea** [comment]
Hi James,

Looks like the Qpid Proton folks added some link accessors several months ago which might help provide our bridge with some sort of visibility to the link health:

https://github.com/apache/qpid-proton/commit/80f6d0013ee4b639cb7ba16e24b0f5594a0e219c

Cheers,
Frank

**Comment by fquinner** [comment]
Taking a note here to update the qpidBridgeMamaPublisher_setUserCallbacks documentation when this has been implemented and to integrate with onError callbacks.

**Comment by fquinner** [comment]
Started work on this in https://github.com/fquinner/OpenMAMA/tree/issue-65. Qpid proton does not seem to offer any asyncronous callbacks so it's a bit hacky, but the commit note sums up status and todos:

```
 QPID: Added some disconnection handling to qpid.

It's a bit crude because qpid doesn't seem to support callback based
event handling but it'll do the job. It's a 3-strikes system which is
specific to each publisher. If a link is down 3 times in a row, the
publisher's endpoint for the failing URL is removed so it will no
longer attempt to publish to that destination. There is still some
work to do around memory management of the endpoint as it's now a
more complex object.

It was debatable whether to adopt this approach or whether to just
crawl all endpoints for the provided URL. On consideration I thought
that would be both slow and very intolerant of reconnecting clients
so I stuck with the 3-strikes per publisher approach.

There is still some more work to do around

- Memory management of the endpoint now it's a more complex object
- Ensuring that endpoints are only created for p2p transports
- I spotted a crash possibly related to issue #153 when (after the
  third client fails), the first client then disconnecting will
  crash
- Valgrind run
```

**Comment by fquinner** [comment]
Succeeded by #162

**Comment by fquinner** [comment]
Fix provided in #162
